src/main/groovy/org/jenkinsci/gradle/plugins/jpi/DependencyAnalysis.groovy

```diff
@@ -0,0 +1,102 @@
+import org.gradle.api.attributes.Attribute
+import org.gradle.api.attributes.Category
+
+@CompileStatic
+class DependencyAnalysis {
+
+    private class JpiConfigurations {
+        Configuration consumableLibraries
+        Configuration consumablePlugins
+        Configuration resolvablePlugins
+
+        JpiConfigurations(Configuration consumableLibraries,
+                          Configuration consumablePlugins,
+                          Configuration resolvablePlugins) {
+            this.consumableLibraries = consumableLibraries
+            this.consumablePlugins = consumablePlugins
+            this.resolvablePlugins = resolvablePlugins
+        }
+    }
+
+    final Configuration allLibraryDependencies
+
+    private static final Attribute CATEGORY_ATTRIBUTE = Attribute.of(Category.CATEGORY_ATTRIBUTE.name, String)
+    private final List<JpiConfigurations> jpiConfigurations = []
+
+    private DependencyAnalysisResult analysisResult
+
+    DependencyAnalysis(Project project) {
+        this.allLibraryDependencies = project.configurations.detachedConfiguration()
+        this.allLibraryDependencies.withDependencies {
+            // do the analysis when this configuration is resolved
+            analyse()
+        }
+    }
+
+    void registerJpiConfigurations(Configuration consumableLibraries,
+                                   Configuration consumablePlugins,
+                                   Configuration resolvablePlugins) {
+        jpiConfigurations.add(new JpiConfigurations(consumableLibraries, consumablePlugins, resolvablePlugins))
+    }
+
+    DependencyAnalysisResult analyse() {
+        if (analysisResult) {
+            return analysisResult
+        }
+
+        def manifestEntry = new StringBuilder()
+
+        jpiConfigurations.each { confs ->
+            analyseDependencies(confs, allLibraryDependencies, manifestEntry)
+        }
+        analysisResult = new DependencyAnalysisResult(manifestEntry.toString())
+        analysisResult
+    }
+
+    private analyseDependencies(JpiConfigurations configurations,
+                                Configuration allLibraries, StringBuilder manifestEntry) {
+        def optional = configurations.resolvablePlugins.name != JpiPlugin.JENKINS_RUNTIME_CLASSPATH_CONFIGURATION_NAME
+
+        configurations.resolvablePlugins.incoming.resolutionResult.root.dependencies.each { DependencyResult result ->
+            if (result.constraint || !(result instanceof ResolvedDependencyResult)) {
+                return
+            }
+            def selected = ((ResolvedDependencyResult) result).selected
+            def moduleVersion = selected.moduleVersion
+            if (moduleVersion == null) {
+                return
+            }
+            selected.variants.each { variant ->
+                if (variant.attributes.getAttribute(CATEGORY_ATTRIBUTE) != Category.LIBRARY) {
+                    //skip platform dependencies
```

Comment:
Are there any other categories? What's the advantage to doing this over checking if the variant is a platform and skipping?

Reply:
Yes there are other categories (like documentation). But in theory they should not turn up here. Although they can if a dependency explicitly depends on such a variant.

To clarify, I expanded the comment in the code and also made it more fail-safe by also checking the `LIBRARY_ELEMENTS_ATTRIBUTE == JPI` attribute.

Reply:
Yes there are other categories (like documentation). But in theory they should not turn up here. Although they can if a dependency explicitly depends on such a variant.

To clarify, I expanded the comment in the code and also made it more fail-safe by also checking the `LIBRARY_ELEMENTS_ATTRIBUTE == JPI` attribute.